File: desktop/services/overview.go
package services

import "errors"

type OverviewService struct {
	proxy *ProxyClient
}

func NewOverviewService(proxy *ProxyClient) *OverviewService {
	return &OverviewService{proxy: proxy}
}

type OverviewDTO struct {
	PhotoCount    int64 `json:"photoCount"`
	DigitalCount  int64 `json:"digitalCount"`
	FilmCount     int64 `json:"filmCount"`
	AlbumCount    int64 `json:"albumCount"`
	StoryCount    int64 `json:"storyCount"`
	BlogCount     int64 `json:"blogCount"`
	FilmRollCount int64 `json:"filmRollCount"`
	FriendCount   int64 `json:"friendCount"`
	CommentCount  int64 `json:"commentCount"`
	CameraCount   int64 `json:"cameraCount"`
	LensCount     int64 `json:"lensCount"`
	CategoryCount int64 `json:"categoryCount"`

	FeaturedCount int64 `json:"featuredCount"`
	HiddenCount   int64 `json:"hiddenCount"`

	PendingComments  int64 `json:"pendingComments"`
	ApprovedComments int64 `json:"approvedComments"`
	RejectedComments int64 `json:"rejectedComments"`

	TotalSize int64 `json:"totalSize"`

	PublishedAlbums  int64 `json:"publishedAlbums"`
	DraftAlbums      int64 `json:"draftAlbums"`
	PublishedStories int64 `json:"publishedStories"`
	DraftStories     int64 `json:"draftStories"`
	PublishedBlogs   int64 `json:"publishedBlogs"`
	DraftBlogs       int64 `json:"draftBlogs"`

	RecentPhotos  []RecentPhotoDTO `json:"recentPhotos"`
	RecentStories []RecentStoryDTO `json:"recentStories"`
	RecentBlogs   []RecentBlogDTO  `json:"recentBlogs"`

	PhotosThisMonth int64 `json:"photosThisMonth"`
	PhotosThisYear  int64 `json:"photosThisYear"`
}

type RecentPhotoDTO struct {
	ID           string  `json:"id"`
	Title        string  `json:"title"`
	URL          string  `json:"url"`
	ThumbnailURL *string `json:"thumbnailUrl,omitempty"`
	CreatedAt    string  `json:"createdAt"`
}

type RecentStoryDTO struct {
	ID          string `json:"id"`
	Title       string `json:"title"`
	CreatedAt   string `json:"createdAt"`
	IsPublished bool   `json:"isPublished"`
}

type RecentBlogDTO struct {
	ID          string `json:"id"`
	Title       string `json:"title"`
	CreatedAt   string `json:"createdAt"`
	IsPublished bool   `json:"isPublished"`
}

func (s *OverviewService) GetOverview() (*OverviewDTO, error) {
	if s.proxy == nil || !s.proxy.IsReady() {
		return nil, errors.New("未连接到服务器")
	}

	var result OverviewDTO
	if err := s.proxy.GET("/admin/overview", &result); err != nil {
		return nil, err
	}
	return &result, nil
}

File: hono/overview.ts
import 'server-only'

import { Hono } from 'hono'

import { db } from '~/server/lib/db'
import { resolvePhotoUrls } from '~/server/lib/photo-urls'

import { authMiddleware, type AuthVariables } from './middleware/auth'

const overview = new Hono<{ Variables: AuthVariables }>()

overview.get('/admin/overview', authMiddleware, async (c) => {
  try {
    const now = new Date()
    const monthStart = new Date(now.getFullYear(), now.getMonth(), 1)
    const photoActivityYear = now.getUTCFullYear()
    const yearStart = new Date(Date.UTC(photoActivityYear, 0, 1))
    const yearEnd = new Date(Date.UTC(photoActivityYear + 1, 0, 1))

    const [
      digitalCount,
      filmCount,
      albumCount,
      storyCount,
      blogCount,
      filmRollCount,
      friendCount,
      commentCount,
      cameraCount,
      lensCount,
      tagCount,
      featuredCount,
      hiddenCount,
      pendingComments,
      approvedComments,
      rejectedComments,
      publishedAlbums,
      publishedStories,
      publishedBlogs,
      sizeAggregate,
      photosThisMonth,
      recentPhotos,
      recentStories,
      recentBlogs,
      dailyRows,
    ] = await Promise.all([
      db.photo.count({ where: { filmPhoto: { is: null } } }),
      db.photo.count({ where: { filmPhoto: { isNot: null } } }),
      db.album.count(),
      db.story.count(),
      db.blog.count(),
      db.filmRoll.count(),
      db.friendLink.count(),
      db.comment.count(),
      db.camera.count(),
      db.lens.count(),
      db.tag.count(),
      db.photo.count({ where: { isFeatured: true } }),
      db.photo.count({ where: { showFlag: false } }),
      db.comment.count({ where: { status: 'pending' } }),
      db.comment.count({ where: { status: 'approved' } }),
      db.comment.count({ where: { status: 'rejected' } }),
      db.album.count({ where: { isPublished: true } }),
      db.story.count({ where: { isPublished: true } }),
      db.blog.count({ where: { isPublished: true } }),
      db.photo.aggregate({ _sum: { size: true } }),
      db.photo.count({ where: { createdAt: { gte: monthStart } } }),
      db.photo.findMany({
        take: 6,
        orderBy: { createdAt: 'desc' },
        select: {
          id: true,
          title: true,
          path: true,
          thumbPath: true,
          storageProvider: true,
          storageSourceId: true,
          storageUrlType: true,
          createdAt: true,
        },
      }),
      db.story.findMany({
        take: 5,
        orderBy: { createdAt: 'desc' },
        select: { id: true, title: true, createdAt: true, isPublished: true },
      }),
      db.blog.findMany({
        take: 5,
        orderBy: { createdAt: 'desc' },
        select: { id: true, title: true, createdAt: true, isPublished: true },
      }),
      // Prisma stores Photo.createdAt as a UTC timestamp without a time zone.
      db.$queryRaw<{ date: string; count: number }[]>`
        SELECT TO_CHAR("createdAt", 'YYYY-MM-DD') AS date, COUNT(*)::int AS count
        FROM "Photo"
        WHERE "createdAt" >= ${yearStart} AND "createdAt" < ${yearEnd}
        GROUP BY 1
        ORDER BY 1
      `,
    ])

    const recentPhotoDtos = await Promise.all(recentPhotos.map(async (photo) => {
      const { url, thumbnailUrl } = await resolvePhotoUrls(photo)
      return {
        id: photo.id,
        title: photo.title,
        url: url ?? '',
        thumbnailUrl,
        createdAt: photo.createdAt.toISOString(),
      }
    }))

    const dailyPhotos = dailyRows.map(row => ({ date: row.date, count: Number(row.count) }))
    const monthlyPhotos = Array.from({ length: 12 }, () => 0)
    let photosThisYear = 0
    for (const day of dailyPhotos) {
      const month = new Date(`${day.date}T00:00:00.000Z`).getUTCMonth()
      monthlyPhotos[month] += day.count
      photosThisYear += day.count
    }

    return c.json({
      success: true,
      data: {
        photoCount: digitalCount + filmCount,
        digitalCount,
        filmCount,
        albumCount,
        storyCount,
        blogCount,
        filmRollCount,
        friendCount,
        commentCount,
        cameraCount,
        lensCount,
        tagCount,
        featuredCount,
        hiddenCount,
        pendingComments,
        approvedComments,
        rejectedComments,
        totalSize: sizeAggregate._sum.size ?? 0,
        publishedAlbums,
        draftAlbums: albumCount - publishedAlbums,
        publishedStories,
        draftStories: storyCount - publishedStories,
        publishedBlogs,
        draftBlogs: blogCount - publishedBlogs,
        recentPhotos: recentPhotoDtos,
        recentStories: recentStories.map((story) => ({
          ...story,
          createdAt: story.createdAt.toISOString(),
        })),
        recentBlogs: recentBlogs.map((blog) => ({
          ...blog,
          createdAt: blog.createdAt.toISOString(),
        })),
        photosThisMonth,
        photosThisYear,
        monthlyPhotos,
        photoActivityYear,
        photoActivityTimeZone: 'UTC',
        dailyPhotos,
      },
    })
  } catch (error) {
    console.error('Get overview error:', error)
    return c.json({ error: 'Internal server error' }, 500)
  }
})

export default overview
